apis/v1alpha2/gateway_types.go

```diff
@@ -0,0 +1,835 @@
+	TCPProtocolType ProtocolType = "TCP"
+
+	// Accepts UDP packets.
+	UDPProtocolType ProtocolType = "UDP"
+)
+
+// GatewayTLSConfig describes a TLS configuration.
+type GatewayTLSConfig struct {
+	// Mode defines the TLS behavior for the TLS session initiated by the client.
+	// There are two possible modes:
+	// - Terminate: The TLS session between the downstream client
+	//   and the Gateway is terminated at the Gateway. This mode requires
+	//   certificateRef to be set.
+	// - Passthrough: The TLS session is NOT terminated by the Gateway. This
+	//   implies that the Gateway can't decipher the TLS stream except for
+	//   the ClientHello message of the TLS protocol.
+	//   CertificateRef field is ignored in this mode.
+	//
+	// Support: Core
+	//
+	// +optional
+	// +kubebuilder:default=Terminate
+	Mode *TLSModeType `json:"mode,omitempty"`
+
+	// CertificateRef is a reference to a Kubernetes object that contains a TLS
+	// certificate and private key. This certificate is used to establish a TLS
+	// handshake for requests that match the hostname of the associated
+	// listener.
+	//
+	// References to a resource in different namespace are invalid UNLESS there
+	// is a ReferencePolicy in the target namespace that allows the certificate
+	// to be attached. If a ReferencePolicy does not allow this reference, the
+	// "ResolvedRefs" condition MUST be set to false for this listener with the
+	// "InvalidCertificateRef" reason.
+	//
+	// This field is required when mode is set to "Terminate" (default) and
+	// optional otherwise.
+	//
+	// CertificateRef can reference a standard Kubernetes resource, i.e. Secret,
+	// or an implementation-specific custom resource.
+	//
+	// Support: Core (Kubernetes Secrets)
+	//
+	// Support: Implementation-specific (Other resource types)
+	//
+	// +optional
+	CertificateRef *ObjectReference `json:"certificateRef,omitempty"`
+
+	// Options are a list of key/value pairs to give extended options
+	// to the provider.
+	//
+	// There variation among providers as to how ciphersuites are
+	// expressed. If there is a common subset for expressing ciphers
+	// then it will make sense to loft that as a core API
+	// construct.
+	//
+	// Support: Implementation-specific
+	//
+	// +optional
+	Options map[string]string `json:"options,omitempty"`
+}
+
+// TLSModeType type defines how a Gateway handles TLS sessions.
+//
+// +kubebuilder:validation:Enum=Terminate;Passthrough
+type TLSModeType string
+
+const (
+	// In this mode, TLS session between the downstream client
+	// and the Gateway is terminated at the Gateway.
+	TLSModeTerminate TLSModeType = "Terminate"
+	// In this mode, the TLS session is NOT terminated by the Gateway. This
+	// implies that the Gateway can't decipher the TLS stream except for
+	// the ClientHello message of the TLS protocol.
+	TLSModePassthrough TLSModeType = "Passthrough"
+)
+
+// ListenerRoutes defines which Routes may be attached to this Listener.
+type ListenerRoutes struct {
```

Comment:
This is name is misnomer. They are `Allowed` routes not `Actual` routes.

Reply:
To clarify, you're suggesting a name more like `ListenerAllowedRoutes`?

Reply:
Yes

Reply:
Will be fixed by #796